include/list.h

```diff
@@ -1,27 +1,179 @@
-#ifndef LIST_H
-#define LIST_H
-
-#include <stdbool.h>
-
-typedef struct List
-{
-    void** _array;
-    int size;
-    int allocated_size;
-} List;
-
-List *list_new(int init_size);
-void list_destroy(List **list);
-bool list_append(List *list, void *value);
-bool list_remove_by_idx(List *list, int index);
-void* list_get(List *list, int index);
-int list_get_size(List *list);
-bool list_remove_by_value(List *list, void *value);
-bool list_exists(List *list, void *value);
-
-bool int_list_append(List *list, intptr_t value);
-intptr_t int_list_get(List *list, int index);
-bool int_list_remove_by_value(List *list, intptr_t value);
-bool int_list_exists(List *list, intptr_t value);
-
-#endif
+/** @file list.h
+ *
+ *  @brief A doubly-linked list
+ *
+ *  List Implementation
+ *  ===================
+ *
+ *  - This @ref List operates as a linked list @ref ListNodes. It operates as a regular doubly-linked list
+ *  but doesn't allocate memory and rather gets @ref ListNodes from a pool.
+ */ 
+#ifndef LIST_H
+#define LIST_H
+
+#include <stdbool.h>
+
+#define MAX_LIST_NODES 128
+
+typedef struct ListNode ListNode;
+
+/**
+ * @brief A single entry in a @ref List
+ */
+struct ListNode
+{
+    /**
+     * @brief The previous @ref ListNode in the associated @ref List, NULL if at the `head` of the list
+     */
+    ListNode* prev;
+
+    /**
+     * @brief The next @ref ListNode in the associated @ref List, NULL if at the `tail` of the list
+     */
+    ListNode* next;
+
+    /**
+     * @brief Pointer to generic data stored in this node
+     */
+    void* data;
+};
+
+/**
+ * @brief A doubly-linked list
+ */
+typedef struct List
+{
+    /**
+     * @brief The first entry in the list
+     */
+    ListNode* head;
+
+    /**
+     * @brief The last entry in the list
+     */
+    ListNode* tail;
+
+    /**
+     * @brief Number of elements in list
+     */
+    int len;
+} List;
+
+/**
+ * @brief An iterator into a list
+ */
+typedef struct
+{
+    /**
+     * @brief A pointer to the @ref List this is iterating through
+     */
+    const List* list;
+
+    /**
+     * @brief The current node in the list
+     */
+    ListNode* current_node;
+} ListItr;
+
+/**
+ * Create a new list.
+ *
```

Comment:
```suggestion
 *
 * The list must be freed with @ref list_destroy
 * While memory is not allocated for the list itself, the list does allocate memory for each element.
 *
```

Add some documentation instructing the list must freed.